lib/charms/traefik_route_k8s/v0/traefik_route.py

```diff
@@ -174,14 +195,33 @@ class TraefikRouteRequirer(Object):
     The TraefikRouteRequirer provides api to store this config in the
     application databag.
     """
+
     on = TraefikRouteRequirerEvents()
+    _stored = StoredState()
 
-    def __init__(self, charm: CharmBase, relation: Relation,
-                 relation_name: str = 'traefik-route'):
+    def __init__(self, charm: CharmBase, relation: Relation, relation_name: str = "traefik-route"):
         super(TraefikRouteRequirer, self).__init__(charm, relation_name)
+        self._stored.set_default(external_host=None)
+
         self._charm = charm
         self._relation = relation
 
+        self.framework.observe(
+            self._charm.on[relation_name].relation_changed, self._on_relation_changed
+        )
+
+    @property
+    def external_host(self) -> str:
+        """Return the external host set by Traefik, if any."""
+        return self._stored.external_host or ""
+
+    def _on_relation_changed(self, event: RelationEvent) -> None:
+        """Update StoredState with external_host and other information from Traefik."""
+        if self._charm.unit.is_leader():
+            external_host = event.relation.data[event.app].get("external_host", "")
+            self._stored.external_host = external_host or self._stored.external_host
+            self.on.ready.emit(event.relation)
```

Comment:
Again I'm not really sure why we are using store state here. Why not just read it from relation data every time?

Reply:
See above, and in case there's some possibility of the API later changing, where trying to retrieve an empty key would "unset" the external host. If that scenario ever happens, this provides a safety net and/or clear means of migration.

Also, again, because without a comparison to some existing state, we push responsibility for "is this external host really different? Should I reconfigure/restart myself?" down to the charm, and then the _charm_ must implement its own `StoredState` anyway, or keep it in peer data, or generate config files/layers and compare them to what's running to decide.

At some point, we need to decide where the real vs desired state is going to be, and keeping it in the library is far less housekeeping/paperwork for users.